elements/lisk-p2p/src/directory/peer_book.ts

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright © 2019 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+
+import { P2PDiscoveredPeerInfo, P2PPeerInfo } from '../p2p_types';
+import { NewPeerConfig, NewPeers } from './new_peers';
+import { TriedPeerConfig, TriedPeers } from './tried_peers';
+
+export interface PeerBookConfig {
+	readonly newPeerConfig?: NewPeerConfig;
+	readonly triedPeerConfig?: TriedPeerConfig;
+	readonly secret: number;
+}
+
+export class PeerBook {
+	private readonly _bannedPeers: ReadonlyArray<P2PDiscoveredPeerInfo>;
+	private readonly _newPeers: NewPeers;
+	private readonly _triedPeers: TriedPeers;
+	public constructor({
+		newPeerConfig,
+		triedPeerConfig,
+		secret,
+	}: PeerBookConfig) {
+		this._newPeers = new NewPeers(newPeerConfig ? newPeerConfig : { secret });
+		this._triedPeers = new TriedPeers(
+			triedPeerConfig ? triedPeerConfig : { secret },
+		);
+		this._bannedPeers = [];
+	}
+
+	public get newPeers(): ReadonlyArray<P2PPeerInfo> {
+		return this._newPeers.getNewPeersList();
+	}
+
+	public get triedPeers(): ReadonlyArray<P2PDiscoveredPeerInfo> {
+		return this._triedPeers.getTriedPeersList();
+	}
+
+	public get bannedPeers(): ReadonlyArray<P2PDiscoveredPeerInfo> {
+		return this._bannedPeers;
+	}
+
+	public getAllPeers(): ReadonlyArray<P2PPeerInfo> {
+		return [...this.newPeers, ...this.triedPeers];
+	}
+
+	// If the peer is completely deleted from both the peer lists then return true
+	public downgradePeer(peerInfo: P2PPeerInfo): boolean {
+		if (this._newPeers.findPeer(peerInfo)) {
+			if (this._newPeers.failedConnectionAction(peerInfo)) {
+				return true;
+			}
+		}
+
+		if (this._triedPeers.findPeer(peerInfo)) {
+			const failed = this._triedPeers.failedConnectionAction(peerInfo);
```

Comment:
We have to downgrade from `tiredPeers` to `newPeers` when there have occurred three failures. Shall we track it in the peer book? Do you wanna try this in a separate issue?

Reply:
The 3 failures are tracked in triedPeer
https://github.com/LiskHQ/lisk-sdk/blob/a634663759c98295c98b6adefcebd33ac2dece48/elements/lisk-p2p/src/directory/tried_peers.ts#L223

once the counter reaches 3 then the function returns true, that means ready to be evicted and when peer_book gets true it simply adds it back to `newPeers`